main/src/library/MutDisjointSet.flix

```diff
@@ -0,0 +1,149 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+pub enum MutDisjointSet[t: Type, r: Eff]({
+    rc = Region[r],
+    forest = MutMap[t, MutDisjointSet.Node[t, r], r]
+})
+
+mod MutDisjointSet {
+
+    pub enum Node[t: Type, r: Eff]({
+        rc = Region[r],
+        value = t,
+        parent = Ref[Option[Node[t, r]], r],
+        rank = Ref[Int32, r]
+    })
+
+    ///
+    /// Returns a new empty `MutDisjointSet` in the region `rc`.
+    ///
+    pub def new(rc: Region[r]): MutDisjointSet[t, r] \ r =
+        MutDisjointSet({ rc = rc, forest = MutMap.new(rc) })
+
+    ///
+    /// Updates `s` with a new disjoint set containing `x`
+    /// if `x` is not already in `s`.
+    ///
+    pub def makeSet(x: t, s: MutDisjointSet[t, r]): Unit \ r with Order[t] = {
+        let MutDisjointSet({ rc, forest }) = s;
+        if (not contains(x, s)) {
+            MutMap.put!(x, singleton(rc, x), forest)
+        } else {
+            ()
+        }
+    }
+
+    ///
+    /// Returns `true` iff `s` contains `x`.
+    ///
+    pub def contains(x: t, s: MutDisjointSet[t, r]): Bool \ r with Order[t] = {
+        let MutDisjointSet({ forest | _ }) = s;
+        match MutMap.get(x, forest) {
+            case Some(_) => true
+            case None    => false
+        }
+    }
+
+    ///
+    /// Returns a new `Node` from `x` in `rc`.
+    ///
+    def singleton(rc: Region[r], x: t): Node[t, r] \ r =
+        Node.Node({ rc = rc, value = x, parent = Ref.new(rc, None), rank = Ref.new(rc, 0) })
+
+    ///
+    /// Updates `s` with collection of disjoint singleton sets containing
+    /// the elements of the iterable `m`.
+    ///
+    pub def makeSets(m: m[t], s: MutDisjointSet[t, r]): Unit \ r with Iterable[m], Order[t] = region rc {
+        Iterable.iterator(rc, m) |> Iterator.forEach(flip(makeSet)(s))
+    }
+
+    ///
+    /// Returns the representative / root of the set that contains `x`.
+    ///
+    pub def find(x: t, s: MutDisjointSet[t, r]): Option[t] \ r with Order[t] =
+        findSet(x, s) |> Option.map(match Node.Node({ value | _ }) -> value)
+
+    ///
+    /// Returns the representative / root `Node` of the set that contains `x`.
+    ///
+    def findSet(x: t, s: MutDisjointSet[t, r]): Option[Node[t, r]] \ r with Order[t] = {
+        def findRoot(y) = {
+            let Node.Node({ parent | _ }) = y;
+            match Ref.get(parent) {
+                case Some(p) => {
+                    let root = findRoot(p);
+                    Ref.put(Some(root), parent);
+                    root
+                }
+                case None => y
+            }
+        };
+        let MutDisjointSet({ forest | _ }) = s;
+        MutMap.get(x, forest)
+        |> Option.map(findRoot)
+    }
+
+    ///
+    /// Merges the sets that contain `x` and `y` if both `x` and `y` are in `s`.
+    /// If `x` and `y` have the same rank, then `y` becomes the new parent of `x`.
```

Comment:
Perhaps some comments on ranking and linking with some URLs?

Reply:
Do we do that anywhere else?

Reply:
Well, we are trying to improve our documentation. And yes we do sometimes give references.